### DISCUSSION TITLE
How to build a custom PrismJS as part of a site build?

### DISCUSSION THREAD

**Discussion by Zinfidel** [PrismJS/prism#3067] [Q&A] [answered]
I want to start off by saying that I am not a webdev person at all, and I am deep into territory I understand little about. I have implemented a custom language definition that I'd like to contribute to a site that uses Prism. Let's assume for the moment that it's not possible to get it accepted as an official addition to the Prism website (more on that later).

Right now, I have managed to stumble my way through the build process and have discovered that, after running `npm run build`, that I can run `npm run start` to get a local webserver with my custom language as part of the list. I include it, click "download", and I get the `prism.js` and `prism.css` files that are the deliverables I need for a PR. This is not ideal though, because if the site owners would like to later customize PrismJS, they either need to perform this process on their own, or lose my customizations.

So, my general question: is there a way to somehow automate the process of `npm run build` > `npm run start` > Click Download > get files? I don't necessarily need to understand the process, I'm asking so that I can hopefully include this kind of information for the webmasters to make the PR more palatable and easy.

On getting my lang def in officially: It's *possible* but I'm not committed to making it a reality. It's for a weird niche language called AviSynth, the definition isn't very good, and the language is very strange in its usage so a lot of style choices I made are very "opinion-territory".

**Comment by RunDevelopment** [comment, marked as answer]
> is there a way to somehow automate the process of `npm run build` > `npm run start` > Click Download > get files?

No, there isn't. (We do have #2618 but I will close that in the future). 

The download page implements a custom bundler. Programmatically using it or adding languages to it is not possible. 

We are very aware that this is an issue. One of the key features of Prism v2 will be that all of Prism language definitions are compatible with widely-used bundlers. We plan to remove the custom bundler and use a bundler (e.g. rollup) instead for the download. This will also support thrid-party languages, so you will the be able to make a little script that create a custom bundle.

> On getting my lang def in officially

You can also open a language request for AviSynth and say: "I'm not doing a PR but here is a basic language definition for anyone that wants to work on it and make a PR".

**Comment by djbpitt** [comment]
In case anyone else needs to run a local server and has as little node experience as I do, I hope that the following will save others the trial-and-error that I needed to fill in the bits I didn't already know.

First, create and `cd` into a new directory that will hold the local clone, and then run the following:

- `git clone --branch v1.30.0 --depth 1 https://github.com/PrismJS/prism.git prism-v1`
-  `cd prism-v1`
 - `npm ci`
 - `npm run build`
 - `npm run start`

If cloning with ssh, the first command should be:

`git clone --branch v1.30.0 --depth 1 git@github.com:PrismJS/prism.git prism-v1`

What I didn't know, even after reading the discussion above, is that:

1. These instructions are for v1. They shouldn't be expected to run with v2. Forking or cloning the prismjs repo copies v2 by default, so I needed to ask for v1 explicitly.
2. Run `npm ci` before building. This installs dependencies that may not already be present.

I hope the preceding is helpful.
